lib/iris/coords.py

```diff
@@ -1373,33 +1446,26 @@ def from_regular(cls, zeroth, step, count, standard_name=None,
             bounds values will be defined. Defaults to False.
 
         """
-        coord = DimCoord.__new__(cls)
-
-        coord.standard_name = standard_name
-        coord.long_name = long_name
-        coord.var_name = var_name
-        coord.units = units
-        coord.attributes = attributes
-        coord.coord_system = coord_system
-        coord.circular = circular
-
-        points = (zeroth+step) + step*np.arange(count, dtype=np.float32)
+        points = (zeroth + step) + step * np.arange(count, dtype=np.float32)
+        # Spoof `iris.util.is_regular`, which only works on coords (which we
+        # don't yet have).
+        diffs = np.diff(points)
+        if not np.allclose(diffs, np.mean(diffs), rtol=0.001):
+            points = (zeroth + step) + step * np.arange(count,
+                                                        dtype=np.float64)
```

Comment:
Perhaps I should have made it more explicit that that code was temporary to confirm it fixed the problem. Duplicating the code was also simpler in the short term than changing the behaviour of the public function `iris.util.regular_step`. 

> simply a few assignments

So, I am now not duplicating code (at all). The original state of this code was not a few assignments, it was a duplication of the functionality of `Coord.__init__`. When I changed the init this method no longer functioned, because it was duplicated code that was not keeping up with the API changes. This version of this method is more robust because we are using the constructor for constructing the coord rather than making a `new`, empty coord bucket and filling it by duplicating most of the functionality of the constructor. 

Further, the code is now much shorter and easier to follow. It also replicates the style of `from_coord`. And the tests pass, which is an improvement on things before I made these changes.